File: behaviour-tree-test/js/nodes/condition.js
/**
 * A Condition node.
 * This acts as a guard and will succeed or fail immediately based on a board predicate, without moving to the 'RUNNING' state.
 * @param conditionFunction The condition function. 
 */
function Condition(conditionFunction) {
    /**
     * The node state.
     */
    let state = NodeState.READY;
   
    /**
     * Update the node and get whether the node state has changed.
     * @param board The board.
     * @returns Whether the state of this node has changed as part of the update.
     */
    this.update = function(board) {
        // Get the pre-update node state.
        const initialState = state;

        // Call the condition function to determine the state of this node.
        state = !!(board[conditionFunction]()) ? NodeState.SUCCEEDED : NodeState.FAILED;

        // Return whether the state of this node has changed.
        return state !== initialState;
    };

    /**
     * Gets the state of the node.
     */
    this.getState = () => state;

    /**
     * Reset the state of the node.
     */
    this.reset = () => {
        // Reset the state of this node.
        state = NodeState.READY;

        // Reset the child node.
        child.reset();
    };
};

File: Dungeony/core/src/com/dumbpug/mistreevous/node/Condition.java
package com.dumbpug.mistreevous.node;

import com.dumbpug.mistreevous.IBoard;
import com.dumbpug.mistreevous.State;
import com.dumbpug.mistreevous.decorator.Decorators;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * A Condition leaf node.
 * This will succeed or fail immediately based on a board predicate, without moving to the 'RUNNING' state.
 */
public class Condition extends Node {
    /**
     * The condition name.
     */
    private String condition;

    /**
     * Creates a new instance of the Condition class.
     * @param decorators The node decorators.
     * @param condition The condition name.
     */
    public Condition(Decorators decorators, String condition) {
        super(decorators);
        this.condition = condition;
    }

    @Override
    public void onUpdate(IBoard board) {
        try {
            // Attempt to get the relevant board method.
            Method method = board.getClass().getMethod(this.condition);

            // Attempt to call the condition method and grab the result.
            boolean result = (boolean)method.invoke(board);

            // The result of calling the condition method will determine the resulting node state.
            this.setState(result ? State.SUCCEEDED : State.FAILED);
        } catch (NoSuchMethodException e) {
            throw new RuntimeException("cannot call CONDITION method '" + this.condition + "' as it is not defined in the blackboard", e);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
    }

    @Override
    public NodeType getNodeType() {
        return NodeType.CONDITION;
    }
}
